### ISSUE TITLE
Improve Cache Implementation

### ISSUE THREAD

**Issue by leakingtapan** [kubernetes-sigs/aws-ebs-csi-driver#35]
The current cache implement has some issues that could be improved:
* It requires explicit `Deprioritize` each time `GetNext` is called. The only time `Deprioritize` is called is after `GetNext`, we can combine the two methods together to help on maintainability and understanbility and reduce the chance that someone forget to call `Deprioritize`
* The behavior of `GetNext` is nondeterministic, this is caused by [sort.Sort](https://golang.org/src/sort/sort.go#L218) which uses quick sort that is unstable sort. 
* Cache is not invalidated after volume is detached.

Proposal:
* Maintain a set of device names that are current attached
* Calling `Next()` each time attaching a volume. Inside `Next()` call, looping through all the combinations of device names and returns the first one that is not in the cache set. This won't cause any performance issue, since the loop iteration count is very small (in the scale of number of device names, less than 100)
* Calling `Remove(deviceName)` each time detaching a volume. Inside `Remove()` call, remove the device from the device name set.

**Comment by leakingtapan** [comment]
/cc @bertinatto 

Let me know how do you think of the problem and my proposal

**Comment by leakingtapan** [comment]
Ah. Found out that this what the in-tree implements as well: https://github.com/kubernetes/kubernetes/blob/v1.12.2/pkg/cloudprovider/providers/aws/device_allocator.go#L42

**Comment by bertinatto** [comment]
> Ah. Found out that this what the in-tree implements as well

Yes, I decoupled the device management code from k8s an created the package `devicemanager` package (and added tests).

> The behavior of GetNext is nondeterministic, this is caused by sort.Sort which uses quick sort that is unstable sort.

This is also from in-tree. What kind of problem can we have with `sort.Sort`?

 > Cache is not invalidated after volume is detached.

I'm not sure if I'm following. The device release is [deferred](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/cloud/cloud.go#L326) even when it has a taint.

**Comment by gnufied** [comment]
I thought the reason `GetNext`  and `Deprioritize` were separate functions is because we wanted to `Deprioritize` only after device has been successfully attached. There is already another mechanism that prevents concurrent use of same device names between attaches.

`GetNext` is indeed non-deterministic and feel free to change it. At the end of day though, the device allocator is a band-aid that exists solely because we have to supply device names on attach. The real fix should be in AWS API somewhere IMO.
